Micrometer metrics

injectives opened the pull request.

shakuzen committed:
[WIP] Micrometer metrics instrumentation

This is a work-in-progress attempt at instrumenting the driver with Micrometer as an optional alternative metrics implementation to the existing internal one.

meistermeier committed:
Bring it to life.

injectives committed:
Update Micrometer metrics

This update includes general improvements, refactorings and tests.

injectives committed:
Merge branch '5.0' into micrometer-metrics

injectives commented:
@shakuzen, big thanks for your effort on this. 👍 
We were wondering if you could have a look at this update and let us know if you are happy to have it merged please?
It would be great if you could also sign the [CLA](https://neo4j.com/developer/cla/) agreement.

shakuzen commented:
Thank you for the mention and for getting this work in shape for merging. I am checking internally about the CLA and will get back to you.

michael-simons commented on driver/src/main/java/org/neo4j/driver/Config.java:
```diff
@@ -95,6 +99,7 @@
     private final boolean isMetricsEnabled;
     private final int eventLoopThreads;
     private final String userAgent;
+    private final MeterRegistry meterRegistry;
```
Are we sure we can have the dependency as optional when we have direct imports / usages in our config?
I bet that will blow up, especially in Graal.

michael-simons reviewed (changes requested):
Can we rethink the optional dependency and go for provided?
Also I'm sceptic whether it will work as expected when Micrometer classes are on the `Config` class itself depending on the classloader used.

Options I see is using the Java service loader (define or use the existing metrics interface, and try to load it via build Java service loader, find either default or micrometer implementation) or something like I did here https://github.com/neo4j-contrib/cypher-dsl/blob/7c8ba2d4892c2933041fe7343d34c7f0a4b26f43/neo4j-cypher-dsl/src/main/java/org/neo4j/cypherdsl/core/Cypher.java#L1113 (this lets you use QueryDSL with CypherDSL).

I think the most suitable approach here would be service loader, tbh.

michael-simons committed:
Make MetricsAdapter a supported, public api.

- Set scope of micrometer to provided
- Rename MetricsProvider to MetricsAdapter (it does not provide metrics, the driver does. It adapts the drivers metrics to something external)
- Make it a public interface
- Remove dependencies to other internal interfaces and classes (such as BoltServerAddress and the pool itself via ServerAddress and intsuppliers)
- Reduce visibility of internal classes as much as possible
- Make sure that adapters are not serialized
- Apply the clock to the internal adapter (an oversight from the previous iteration)

michael-simons committed:
Saving a file before commiting seems to be relevant.

michael-simons committed:
Not messing up the file while saven even more so.

michael-simons committed:
Make interfaces internal again, provide an enum to select the provider, use global registry by default.

This change keeps the existing serialization feature of config correct.
It will work for everyone just using micrometer as is with the global registry.
if there is a need to select the registriy, we can work on that later.

meistermeier commented on driver/src/main/java/org/neo4j/driver/Config.java:
```diff
@@ -714,7 +723,37 @@ public ConfigBuilder withDriverMetrics()
+        {
+            if ( !enabled )
+            {
+                this.metricsAdapter = MetricsAdapter.DEV_NULL;
```
```suggestion
                return withMetricsAdapter( MetricsAdapter.DEV_NULL );
```

meistermeier commented on driver/src/main/java/org/neo4j/driver/Config.java:
```diff
@@ -714,7 +723,37 @@ public ConfigBuilder withDriverMetrics()
+            }
+            else if ( this.metricsAdapter == null || this.metricsAdapter == MetricsAdapter.DEV_NULL )
+            {
+                this.metricsAdapter = MetricsAdapter.DEFAULT;
```
```suggestion
                return withMetricsAdapter( MetricsAdapter.DEFAULT );
```

meistermeier commented on driver/src/main/java/org/neo4j/driver/ConnectionPoolMetrics.java:
```diff
@@ -59,7 +60,7 @@
      * A counter to record how many connections have been successfully created with this pool since the pool is created.
      * This number increases every time when a connection is successfully created.
-     * @return The amount of connections have ever been created by this pool.
+     * @return The amount of connections has ever been created by this pool.
```
```suggestion
     * @return The amount of connections that has ever been created by this pool.
```

meistermeier commented on driver/src/test/java/org/neo4j/driver/ConfigTest.java:
```diff
@@ -370,7 +400,7 @@ void shouldErrorWithInvalidUserAgent()
         {
             Config config = Config.builder()
-                    .withMaxConnectionPoolSize( 123 )
+                                  .withMaxConnectionPoolSize( 123 )
```
```suggestion
                    .withMaxConnectionPoolSize( 123 )
```

injectives replied:
Updated formatting in https://github.com/neo4j/neo4j-java-driver/pull/1137/commits/d942e0deb9a5f26a6dda13fe2258723aea80980d.

meistermeier reviewed (commented):
Do we have to mention somewhere which version(s) of Micrometer is/are supported?
Even though I like the style and would do the same I suggest a refactoring (**not** part of this PR) away from `/dev/null` (there is also logging) naming to something different, more understandable for non-neerdy code readers.

injectives committed:
Move micrometer-core optional tag to dependency declaration

injectives committed:
Update Config javadoc and remove imports of internal metrics classes

injectives committed:
Update withMetricsEnabled(boolean) implementation

injectives committed:
Enable ConfigBuilder.withMetricsAdapter documentation

injectives committed:
Update driver/src/main/java/org/neo4j/driver/ConnectionPoolMetrics.java

Co-authored-by: Gerrit Meier <meistermeier@gmail.com>

injectives committed:
Update ConnectionPoolMetrics documentation

injectives committed:
Update MetricsAdapter documentation

injectives committed:
Update formatting in ConfigTest

injectives committed:
Update wording

meistermeier commented on driver/src/main/java/org/neo4j/driver/internal/metrics/DevNullMetricsListener.java:
```diff
@@ -0,0 +1,100 @@
+    @Override
+    public String toString()
+    {
+        return "Driver metrics are not available when they are not enabled.";
```
```suggestion
        return "Driver metrics are not available if they are not enabled.";
```
?

injectives committed:
Updated wording

injectives committed:
Formatting

injectives committed:
Update driver/src/main/java/org/neo4j/driver/internal/metrics/DevNullMetricsListener.java

Co-authored-by: Gerrit Meier <meistermeier@gmail.com>

meistermeier commented on driver/src/main/java/org/neo4j/driver/internal/metrics/MicrometerConnectionPoolMetrics.java:
```diff
@@ -0,0 +1,277 @@
+                       totalAcquisitionTime(), totalConnectionTime(), totalInUseTime(), totalInUseCount() );
+    }
+
+    private long toLong( Counter counter )
```
```suggestion
    private long count( Counter counter )
```
or something similar, more "business logic" oriented?

meistermeier reviewed (approved):
Minor renaming. But if you think that it is better as it is right now, feel free to ignore it.

injectives committed:
Update driver/src/main/java/org/neo4j/driver/internal/metrics/MicrometerConnectionPoolMetrics.java

Co-authored-by: Gerrit Meier <meistermeier@gmail.com>

shakuzen commented on driver/src/main/java/org/neo4j/driver/internal/metrics/MicrometerConnectionPoolMetrics.java:
```diff
@@ -0,0 +1,277 @@
+    public void afterCreated( ListenerEvent<?> connEvent )
+    {
+        creating.decrementAndGet();
+        created.increment();
```
Timers include a count, so a separate Counter is not needed if it counts the same events being timed.

injectives replied:
Good point, thanks.

shakuzen commented on driver/src/main/java/org/neo4j/driver/internal/metrics/MicrometerConnectionPoolMetrics.java:
```diff
@@ -0,0 +1,277 @@
+    @Override
+    public void afterAcquiredOrCreated( ListenerEvent<?> acquireEvent )
+    {
+        acquired.increment();
```
The acquired count can be obtained from the Timer for the same thing.

injectives replied:
Thank you.

shakuzen commented on driver/src/main/java/org/neo4j/driver/internal/metrics/MicrometerConnectionPoolMetrics.java:
```diff
@@ -0,0 +1,277 @@
+    @Override
+    public void released( ListenerEvent<?> inUseEvent )
+    {
+        released.increment();
```
Likewise, this can be derived from the Timer.

injectives replied:
Thanks.

shakuzen reviewed (commented):
This is looking very nice. I see the `MicrometerMetricsProvider` can take an arbitrary `MeterRegistry` instance - this is good. How does a user configure things so a specific `MeterRegistry` is used?

injectives committed:
Merge branch '5.0' into micrometer-metrics

injectives committed:
Merge branch '5.0' into micrometer-metrics

injectives committed:
Fix compilation error

injectives committed:
Update failing test

injectives committed:
Fix GetConnectionPoolMetrics access

injectives committed:
Delete redundant counters based on review feedback

injectives commented:
> This is looking very nice. I see the `MicrometerMetricsProvider` can take an arbitrary `MeterRegistry` instance - this is good. How does a user configure things so a specific `MeterRegistry` is used?

Conceptually, it can take any instance. However, at this iteration and for initial period we are only going to support the global registry.

@michael-simons, please correct me if I am wrong.

injectives committed:
Merge branch '5.0' into micrometer-metrics

michael-simons reviewed (approved):
Looks good from my site, and yes, @injectives that was the intention: If this feature gets traction we shall revaluate custom registries.
